// @kwilteam/kwil-js/dist/funder/funder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Funder = void 0;
const enums_1 = require("../core/enums");
const payloadTx_1 = require("../transaction/payloadTx");
const serial_1 = require("../utils/serial");
class Funder {
    constructor(kwil, funderClient, chainId) {
        this.kwil = kwil;
        this.funderClient = funderClient;
        this.chainId = chainId;
    }
    async transfer(payload, signer, synchronous) {
        if (!payload.keyType) {
            payload.keyType = enums_1.AccountKeyType.SECP256K1;
        }
        if (payload.to instanceof Uint8Array) {
            payload.to = (0, serial_1.bytesToHex)(payload.to);
        }
        const txPayload = {
            to: {
                identifier: payload.to,
                key_type: payload.keyType,
            },
            amount: payload.amount.toString(),
        };
        const transaction = await payloadTx_1.PayloadTx.createTx(this.kwil, {
            chainId: this.chainId,
            description: payload.description || '',
            payload: txPayload,
            payloadType: enums_1.PayloadType.TRANSFER,
            identifier: signer.identifier,
            signer: signer.signer,
            signatureType: signer.signatureType,
        }).buildTx();
        return await this.funderClient.broadcastClient(transaction, synchronous ? enums_1.BroadcastSyncType.COMMIT : enums_1.BroadcastSyncType.SYNC);
    }
}
exports.Funder = Funder;
