// @kwilteam/kwil-js/dist/core/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessModifier = exports.AccountKeyType = exports.AccountStatus = exports.AuthErrorCodes = exports.BroadcastSyncType = exports.AuthenticationMode = exports.EnvironmentType = exports.BytesEncodingStatus = exports.SerializationType = exports.PayloadType = exports.EncodingType = exports.IndexType = exports.AttributeType = exports.VarType = void 0;
/**
 * VarType is the type of the data in the database.
 *
 * Although Kwil supports text, int, bool, blob, uuid, uint256, decimal, and null types, kwil-js only supports text, int, bool, decimal, and null. If you need to send blob, uuid, or uint256 types, you should send them as a javascript string.
 *
 */
var VarType;
(function (VarType) {
    VarType["UUID"] = "uuid";
    VarType["TEXT"] = "text";
    VarType["INT8"] = "int8";
    VarType["BOOL"] = "bool";
    VarType["NUMERIC"] = "numeric";
    VarType["NULL"] = "null";
    VarType["BYTEA"] = "bytea";
    VarType["UNKNOWN"] = "unknown";
})(VarType || (exports.VarType = VarType = {}));
var AttributeType;
(function (AttributeType) {
    AttributeType["INVALID_TYPE"] = "";
    AttributeType["PRIMARY_KEY"] = "PRIMARY_KEY";
    AttributeType["UNIQUE"] = "UNIQUE";
    AttributeType["NOT_NULL"] = "NOT_NULL";
    AttributeType["DEFAULT"] = "DEFAULT";
    AttributeType["MIN"] = "MIN";
    AttributeType["MAX"] = "MAX";
    AttributeType["MIN_LENGTH"] = "MIN_LENGTH";
    AttributeType["MAX_LENGTH"] = "MAX_LENGTH";
})(AttributeType || (exports.AttributeType = AttributeType = {}));
var IndexType;
(function (IndexType) {
    IndexType["INVALID_INDEX_TYPE"] = "";
    IndexType["BTREE"] = "BTREE";
    IndexType["UNIQUE_BTREE"] = "UNIQUE_BTREE";
})(IndexType || (exports.IndexType = IndexType = {}));
var EncodingType;
(function (EncodingType) {
    EncodingType[EncodingType["INVALID_ENCODING_TYPE"] = 0] = "INVALID_ENCODING_TYPE";
    EncodingType[EncodingType["RLP_ENCODING"] = 1] = "RLP_ENCODING";
})(EncodingType || (exports.EncodingType = EncodingType = {}));
var PayloadType;
(function (PayloadType) {
    PayloadType["INVALID_PAYLOAD_TYPE"] = "invalid";
    PayloadType["EXECUTE_ACTION"] = "execute";
    PayloadType["CALL_ACTION"] = "call_action";
    PayloadType["TRANSFER"] = "transfer";
    PayloadType["RAW_STATEMENT"] = "raw_statement";
})(PayloadType || (exports.PayloadType = PayloadType = {}));
var SerializationType;
(function (SerializationType) {
    SerializationType["INVALID_SERIALIZATION_TYPE"] = "invalid";
    SerializationType["SIGNED_MSG_CONCAT"] = "concat";
    SerializationType["SIGNED_MSG_EIP712"] = "eip712";
})(SerializationType || (exports.SerializationType = SerializationType = {}));
var BytesEncodingStatus;
(function (BytesEncodingStatus) {
    BytesEncodingStatus["INVALID_ENCODING_STATUS"] = "invalid";
    BytesEncodingStatus["BASE64_ENCODED"] = "base64_encoded";
    BytesEncodingStatus["HEX_ENCODED"] = "hex_encoded";
    BytesEncodingStatus["UINT8_ENCODED"] = "uint8_encoded";
})(BytesEncodingStatus || (exports.BytesEncodingStatus = BytesEncodingStatus = {}));
var EnvironmentType;
(function (EnvironmentType) {
    EnvironmentType["BROWSER"] = "browser";
    EnvironmentType["NODE"] = "node";
})(EnvironmentType || (exports.EnvironmentType = EnvironmentType = {}));
var AuthenticationMode;
(function (AuthenticationMode) {
    AuthenticationMode["PRIVATE"] = "private";
    AuthenticationMode["OPEN"] = "open";
})(AuthenticationMode || (exports.AuthenticationMode = AuthenticationMode = {}));
var BroadcastSyncType;
(function (BroadcastSyncType) {
    BroadcastSyncType[BroadcastSyncType["SYNC"] = 0] = "SYNC";
    BroadcastSyncType[BroadcastSyncType["COMMIT"] = 1] = "COMMIT";
})(BroadcastSyncType || (exports.BroadcastSyncType = BroadcastSyncType = {}));
var AuthErrorCodes;
(function (AuthErrorCodes) {
    AuthErrorCodes[AuthErrorCodes["PRIVATE_MODE"] = -1001] = "PRIVATE_MODE";
    AuthErrorCodes[AuthErrorCodes["KGW_MODE"] = -901] = "KGW_MODE";
})(AuthErrorCodes || (exports.AuthErrorCodes = AuthErrorCodes = {}));
// For checking the unconfirmed nonce
var AccountStatus;
(function (AccountStatus) {
    // returns the latest confirmed nonce
    AccountStatus[AccountStatus["LATEST"] = 0] = "LATEST";
    // returns the latest unconfirmed nonce
    AccountStatus[AccountStatus["PENDING"] = 1] = "PENDING";
})(AccountStatus || (exports.AccountStatus = AccountStatus = {}));
var AccountKeyType;
(function (AccountKeyType) {
    // Eth
    AccountKeyType["SECP256K1"] = "secp256k1";
    // ED25519
    AccountKeyType["ED25519"] = "ed25519";
})(AccountKeyType || (exports.AccountKeyType = AccountKeyType = {}));
var AccessModifier;
(function (AccessModifier) {
    AccessModifier["PUBLIC"] = "PUBLIC";
    AccessModifier["PRIVATE"] = "PRIVATE";
    AccessModifier["VIEW"] = "VIEW";
})(AccessModifier || (exports.AccessModifier = AccessModifier = {}));
