// @kwilteam/kwil-js/dist/core/jsonrpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSONRPCMethod = void 0;
var JSONRPCMethod;
(function (JSONRPCMethod) {
    // TODO: Should we implement version? It hasn't historically existed on Kwil-JS, but i notice it in kwil-db.
    JSONRPCMethod["METHOD_HEALTH"] = "user.health";
    JSONRPCMethod["METHOD_PING"] = "user.ping";
    JSONRPCMethod["METHOD_CHAIN_INFO"] = "user.chain_info";
    JSONRPCMethod["METHOD_ACCOUNT"] = "user.account";
    JSONRPCMethod["METHOD_BROADCAST"] = "user.broadcast";
    JSONRPCMethod["METHOD_CALL"] = "user.call";
    JSONRPCMethod["METHOD_DATABASES"] = "user.databases";
    JSONRPCMethod["METHOD_PRICE"] = "user.estimate_price";
    JSONRPCMethod["METHOD_QUERY"] = "user.query";
    JSONRPCMethod["METHOD_TX_QUERY"] = "user.tx_query";
    JSONRPCMethod["METHOD_SCHEMA"] = "user.schema";
    JSONRPCMethod["METHOD_KGW_PARAM"] = "kgw.authn_param";
    JSONRPCMethod["METHOD_KGW_AUTHN"] = "kgw.authn";
    JSONRPCMethod["METHOD_KGW_LOGOUT"] = "kgw.logout";
    JSONRPCMethod["METHOD_CHALLENGE"] = "user.challenge";
})(JSONRPCMethod || (exports.JSONRPCMethod = JSONRPCMethod = {}));
