// @kwilteam/kwil-js/dist/core/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Msg = exports.BaseMessage = void 0;
const signature_1 = require("./signature");
/**
 * `BaseMessage` is the bass class for the payload structure for a request to the Kwil `call` GRPC endpoint {@link https://github.com/kwilteam/proto/blob/main/kwil/tx/v1/call.proto}.
 *
 * Bytes in the message can be typed to be either base64 encoded or Uint8Array. Uint8Array should be used when building the message within the SDK, and base64 should be used for the final message to be send over GRPC.
 *
 * @template {BytesEncodingStatus.BASE64_ENCODED | BytesEncodingStatus.UINT8_ENCODED} T - The type of bytes in the message. Can be either base64 encoded or Uint8Array.
 * @implements {MsgData<T>} - The message data interface.
 */
class BaseMessage {
    constructor(data) {
        // create a basic template of msg. Null values are used to be compatible with both types in PayloadBytesTypes.
        this.data = data || {
            body: {
                payload: null,
                challenge: '',
            },
            auth_type: signature_1.SignatureType.SECP256K1_PERSONAL,
            sender: null,
            signature: null,
        };
    }
    get body() {
        return this.data.body;
    }
    get auth_type() {
        return this.data.auth_type;
    }
    get sender() {
        return this.data.sender;
    }
    get signature() {
        return this.data.signature;
    }
}
exports.BaseMessage = BaseMessage;
var Msg;
(function (Msg) {
    /**
     * Creates a new instance of the `BaseMessage` class.
     *
     * Bytes in the message can be typed to be either base64 encoded or Uint8Array. Uint8Array should be used when building the message within the SDK, and base64 should be used for the final message to be send over GRPC.
     *
     * @template {BytesEncodingStatus.BASE64_ENCODED | BytesEncodingStatus.UINT8_ENCODED} T - The type of bytes in the message. Can be either base64 encoded or Uint8Array.
     * @param {(msg: MsgData<T>) => void} configure - A callback function that takes in a `MsgData` object and sets fields on it.
     * @returns {BaseMessage<T>} - A new instance of the `BaseMessage` class.
     */
    function create(configure) {
        // create a basic template of msg. Null values are used to be compatible with both types in PayloadBytesTypes.
        const msg = {
            body: {
                payload: null,
                challenge: '',
            },
            auth_type: signature_1.SignatureType.SECP256K1_PERSONAL,
            sender: null,
            signature: null,
        };
        // Pass the 'msg' object to the 'configure' function allowing external modification of its propoerties before instantiation of BaseMessage.
        configure(msg);
        return new BaseMessage(msg);
    }
    Msg.create = create;
    /**
     * Copies an existing instance of the `BaseMessage` class and modifies certain fields.
     *
     * Bytes in the message can be typed to be either base64 encoded or Uint8Array. Uint8Array should be used when building the message within the SDK, and base64 should be used for the final message to be send over GRPC.
     *
     * @template {BytesEncodingStatus.BASE64_ENCODED | BytesEncodingStatus.UINT8_ENCODED} T - The type of bytes in the message. Can be either base64 encoded or Uint8Array.
     * @param {BaseMessage<PayloadBytesTypes>} source - The source message to copy from. It can be using either base64 or Uint8Array bytes.
     * @param {(msg: MsgData<T>) => void} configure - A callback function that takes in a `MsgData` object and sets fields on it.
     * @returns {BaseMessage<T>} - A new instance of the `BaseMessage` class.
     */
    function copy(source, configure) {
        return Msg.create((msg) => {
            // copy all fields from source to msg
            msg.body = source.body;
            msg.auth_type = source.auth_type;
            msg.sender = source.sender;
            msg.signature = source.signature;
            // Pass the 'msg' object to the 'configure' function allowing external modification of its propoerties before instantiation of BaseMessage.
            configure(msg);
        });
    }
    Msg.copy = copy;
})(Msg || (exports.Msg = Msg = {}));
