// @kwilteam/kwil-js/dist/core/tx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Txn = exports.BaseTransaction = void 0;
const signature_1 = require("./signature");
const strings_1 = require("../utils/strings");
const enums_1 = require("./enums");
/**
 * `BaseTransaction` is the base class for a payload structure for a a request to the Kwil `broadcast` GRPC endpoint {@link https://github.com/kwilteam/proto/blob/main/kwil/tx/v1/broadcast.proto}.
 * Bytes in the transaction can be typed to be either base64 encoded or Uint8Array. Uint8Array should be used when building the transaction within the SDK, and base64 should be used for the final transaction to be send over GRPC.
 *
 * @template {BytesEncodingStatus.BASE64_ENCODED | BytesEncodingStatus.UINT8_ENCODED} T - The type of bytes in the transaction. Can be either base64 encoded or Uint8Array.
 * @implements {TxnData<T>} - The transaction data interface.
 */
class BaseTransaction {
    constructor(data) {
        // create basic template of tx. Null values are used to be compatible with both types in PayloadBytesTypes.
        this.data = data || {
            signature: {
                sig: null,
                type: signature_1.SignatureType.SIGNATURE_TYPE_INVALID
            },
            body: {
                desc: '',
                payload: null,
                type: enums_1.PayloadType.INVALID_PAYLOAD_TYPE,
                fee: null,
                nonce: null,
                chain_id: ''
            },
            sender: null,
            serialization: enums_1.SerializationType.SIGNED_MSG_CONCAT
        };
    }
    get txData() {
        return this.data;
    }
    isSigned() {
        return !strings_1.strings.isNilOrEmpty(this.data.signature.sig);
    }
    get signature() {
        return this.data.signature;
    }
    get sender() {
        return this.data.sender;
    }
    get body() {
        return this.data.body;
    }
    get serialization() {
        return this.data.serialization;
    }
}
exports.BaseTransaction = BaseTransaction;
var Txn;
(function (Txn) {
    /**
     * Creates a new instance of the `BaseTransaction` class.
     * Bytes in the transaction can be typed to be either base64 encoded or Uint8Array. Uint8Array should be used when building the transaction within the SDK, and base64 should be used for the final transaction to be send over GRPC.
     *
     * @template {BytesEncodingStatus.BASE64_ENCODED | BytesEncodingStatus.UINT8_ENCODED} T - The type of bytes in the transaction. Can be either base64 encoded or Uint8Array.
     * @param {(tx: TxnData<T>) => void} configure - A function that takes in a `TxnData` object and sets fields on it.
     * @returns {BaseTransaction<T>} - A new instance of the `BaseTransaction` class.
     */
    function create(configure) {
        // create basic template of tx. Null values are used to be compatible with both types in PayloadBytesTypes.
        const tx = {
            signature: {
                sig: null,
                type: signature_1.SignatureType.SECP256K1_PERSONAL
            },
            body: {
                desc: '',
                payload: null,
                type: enums_1.PayloadType.INVALID_PAYLOAD_TYPE,
                fee: null,
                nonce: null,
                chain_id: ''
            },
            sender: null,
            serialization: enums_1.SerializationType.SIGNED_MSG_CONCAT
        };
        // Passes the 'tx' object to the 'configure' function allowing external modification of its properties before instantiation of BaseTransaction.
        configure(tx);
        return new BaseTransaction(tx);
    }
    Txn.create = create;
    /**
     * Copies an existing instance of the `BaseTransaction` class and modifies certain fields.
     *
     * Bytes in the transaction can be typed to be either base64 encoded or Uint8Array. Uint8Array should be used when building the transaction within the SDK, and base64 should be used for the final transaction to be send over GRPC.
     *
     * @template {BytesEncodingStatus.BASE64_ENCODED | BytesEncodingStatus.UINT8_ENCODED} T - The type of bytes in the transaction. Can be either base64 encoded or Uint8Array.
     * @param {BaseTransaction<PayloadBytesTypes>} source - The transaction to copy. It can be using either bytes types.
     * @param {(tx: TxnData<T>) => void} configure - A function that takes in a `TxnData` object and sets fields on it.
     * @returns {BaseTransaction<T>} - A new instance of the `BaseTransaction` class.
     */
    function copy(source, configure) {
        return Txn.create((tx) => {
            // copy all fields from source tx to new tx.
            tx.body = source.body;
            tx.signature = source.signature;
            tx.body = source.body;
            tx.sender = source.sender;
            tx.serialization = source.serialization;
            // Passes the 'tx' object to the 'configure' function allowing external modification of its properties before instantiation of BaseTransaction.
            configure(tx);
        });
    }
    Txn.copy = copy;
})(Txn || (exports.Txn = Txn = {}));
