// @kwilteam/kwil-js/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.Utils = exports.KwilSigner = exports.WebKwil = exports.NodeKwil = void 0;
const nodeKwil_1 = require("./client/node/nodeKwil");
Object.defineProperty(exports, "NodeKwil", { enumerable: true, get: function () { return nodeKwil_1.NodeKwil; } });
const webKwil_1 = require("./client/web/webKwil");
Object.defineProperty(exports, "WebKwil", { enumerable: true, get: function () { return webKwil_1.WebKwil; } });
const dbid_1 = require("./utils/dbid");
const action_1 = require("./core/action");
const database_1 = require("./core/database");
const kwilSigner_1 = require("./core/kwilSigner");
Object.defineProperty(exports, "KwilSigner", { enumerable: true, get: function () { return kwilSigner_1.KwilSigner; } });
const client_1 = __importDefault(require("./api_client/client"));
exports.Client = client_1.default;
var Utils;
(function (Utils) {
    /**
     * `ActionInput` class is a utility class for creating action inputs.
     * @deprecated This class is deprecated and will be removed in the next major release.  Please use the `params` instead.
     */
    Utils.ActionInput = action_1.ActionInput;
    /**
     * Generates a unique database identifier (DBID) from the provided owner's public key and a database name.
     * @deprecated This function is deprecated and will be removed in the next major release.
     */
    Utils.generateDBID = dbid_1.generateDBID;
    /**
     * `DataType` holds the different data types that can be asserted as action inputs.
     */
    Utils.DataType = database_1.DataType;
})(Utils || (exports.Utils = Utils = {}));
