// @kwilteam/kwil-js/dist/utils/crypto.js
"use strict";
// noinspection JSPotentiallyInvalidConstructorUsage
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sha256BytesToBytes = exports.sha224BytesToString = exports.sha224StringToString = exports.sha384StringToBytes = exports.sha384BytesToBytes = exports.sha384BytesToString = exports.sha384StringToString = void 0;
const jssha_1 = __importDefault(require("jssha"));
function sha384StringToString(message) {
    // noinspection JSPotentiallyInvalidConstructorUsage
    const shaObj = new jssha_1.default('SHA-384', 'TEXT');
    shaObj.update(message);
    return shaObj.getHash('HEX');
}
exports.sha384StringToString = sha384StringToString;
function sha384BytesToString(message) {
    const shaObj = new jssha_1.default('SHA-384', 'UINT8ARRAY');
    shaObj.update(message);
    return shaObj.getHash('HEX');
}
exports.sha384BytesToString = sha384BytesToString;
function sha384BytesToBytes(message) {
    const shaObj = new jssha_1.default('SHA-384', 'UINT8ARRAY');
    shaObj.update(message);
    return shaObj.getHash('UINT8ARRAY');
}
exports.sha384BytesToBytes = sha384BytesToBytes;
function sha384StringToBytes(message) {
    const shaObj = new jssha_1.default('SHA-384', 'TEXT');
    shaObj.update(message);
    return shaObj.getHash('UINT8ARRAY');
}
exports.sha384StringToBytes = sha384StringToBytes;
function sha224StringToString(message) {
    const shaObj = new jssha_1.default('SHA-224', 'TEXT');
    shaObj.update(message);
    return shaObj.getHash('HEX');
}
exports.sha224StringToString = sha224StringToString;
function sha224BytesToString(message) {
    const shaObj = new jssha_1.default('SHA-224', 'UINT8ARRAY');
    shaObj.update(message);
    return shaObj.getHash('HEX');
}
exports.sha224BytesToString = sha224BytesToString;
function sha256BytesToBytes(message) {
    const shaObj = new jssha_1.default('SHA-256', 'UINT8ARRAY');
    shaObj.update(message);
    return shaObj.getHash('UINT8ARRAY');
}
exports.sha256BytesToBytes = sha256BytesToBytes;
